File: linux-client/downloader.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"time"
)

// DownloadUpdate 下载更新包并验证 hash，返回临时文件路径
func DownloadUpdate(url, filename, expectedHash string) (string, error) {
	log.Printf("[DOWNLOADER] Starting download: %s", url)

	// 1. 生成临时文件路径
	timestamp := fmt.Sprintf("%d", time.Now().Unix())
	tempPath := fmt.Sprintf("/tmp/systemhardening-update-%s.zip", timestamp)

	// 2. 发起 HTTP 请求
	resp, err := http.Get(url)
	if err != nil {
		return "", fmt.Errorf("HTTP request failed: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("download failed: HTTP %d", resp.StatusCode)
	}

	// 3. 创建临时文件
	file, err := os.Create(tempPath)
	if err != nil {
		return "", fmt.Errorf("create temp file failed: %v", err)
	}
	defer file.Close()

	// 4. 计算 MD5 hash
	hasher := md5.New()

	// 5. 写入文件并计算 hash
	written, err := io.Copy(io.MultiWriter(file, hasher), resp.Body)
	if err != nil {
		os.Remove(tempPath)
		return "", fmt.Errorf("download failed: %v", err)
	}

	log.Printf("[DOWNLOADER] Downloaded %d bytes to %s", written, tempPath)

	// 6. 关闭文件
	if err := file.Close(); err != nil {
		os.Remove(tempPath)
		return "", fmt.Errorf("close file failed: %v", err)
	}

	// 7. 验证 hash
	actualHash := hex.EncodeToString(hasher.Sum(nil))
	log.Printf("[DOWNLOADER] MD5 verification: expected=%s, actual=%s", expectedHash, actualHash)

	if actualHash != expectedHash {
		os.Remove(tempPath)
		return "", fmt.Errorf("hash mismatch: expected %s, got %s", expectedHash, actualHash)
	}

	log.Printf("[DOWNLOADER] ✅ MD5 verified successfully")

	return tempPath, nil
}

File: windows-client/src/downloader.rs
//! Windows 客户端下载模块 (阻塞式)

use std::fs;
use std::io::{self, Write};
use std::path::PathBuf;

/// 下载更新包并验证 hash (阻塞式实现)
pub fn download_update(url: &str, _filename: &str, _expected_hash: &str) -> Result<String, String> {
    log::info!("[DOWNLOADER] Starting download: {}", url);
    
    // 1. 生成临时文件路径
    let timestamp = chrono::Local::now().format("%Y%m%d%H%M%S").to_string();
    let temp_dir = PathBuf::from(r"C:\ProgramData\SystemHardening\temp");
    fs::create_dir_all(&temp_dir)
        .map_err(|e| format!("Create temp dir failed: {}", e))?;
    
    let temp_path = temp_dir.join(format!("{}-update-{}.exe", "systemhardening", timestamp));
    
    // 2. 发起 HTTP 请求 (blocking)
    let mut resp = reqwest::blocking::get(url)
        .map_err(|e| format!("HTTP request failed: {}", e))?;
    
    if !resp.status().is_success() {
        return Err(format!("download failed: HTTP {}", resp.status()));
    }
    
    // 3. 创建临时文件
    let mut file = fs::File::create(&temp_path)
        .map_err(|e| format!("Create temp file failed: {}", e))?;
    
    // 4. 写入数据并计算大小
    let downloaded = io::copy(&mut resp, &mut file)
        .map_err(|e| format!("Write to file failed: {}", e))?;
    
    log::info!("[DOWNLOADER] Downloaded {} bytes to: {}", downloaded, temp_path.display());
    
    Ok(temp_path.to_string_lossy().to_string())
}
